python/lsst/daf/butler/core/datasets.py

```diff
@@ -330,6 +330,12 @@ class DatasetRef:
 
     def __init__(self, datasetType, dataId, id=None, run=None, hash=None, components=None):
         assert isinstance(datasetType, DatasetType)
+
+        # Check the universes match if a DataId is provided
+        if isinstance(dataId, DataId):
+            if dataId.dimensions().universe != datasetType.dimensions.universe:
+                raise ValueError(f"Dimension universe mismatch for {dataId} and {datasetType}")
```

Comment:
I think it's safe to assume that the universes match - the only way they wouldn't be is if you've got two Registry instances from different dimension configurations in play, and even then I don't know that it's a serious problem if they don't match.

What is important is that the dimensions themselves do match, i.e. `dataId.dimensions() == datasetType.dimensions` - but this comparison can only safely be done after `datasetType.dimensions` has been "upgraded" to a `DimensionGraph` (`dataId.dimensions()` is always guaranteed to be a `DimensionGraph`).